### ISSUE TITLE
[Question]  Huge performance difference with different n_envs

### ISSUE THREAD

**Issue by Darklanx** [DLR-RM/rl-baselines3-zoo#207]
### Question
I'm running a2c with default parameters on BreakoutNoFrameskip-v4, with two different training scenario, where the only difference is that one uses `n_envs=16` (orange) while the other one set `n_envs=40` (blue). However the performance difference is huge. Is there a particular reason behind this behavior? I thought `n_envs` is more of a parallelization parameter, which shouldn't have such a huge impact on performance. 
![image](https://user-images.githubusercontent.com/26689157/153244495-a7b90c96-5f3f-47eb-90fb-1b46adac635d.png)

### Additional context
config.yml:
```
!!python/object/apply:collections.OrderedDict
- - - ent_coef
    - 0.01
  - - env_wrapper
    - - stable_baselines3.common.atari_wrappers.AtariWrapper
  - - frame_stack
    - 4
  - - n_envs
    - 40 (or 16)
  - - n_timesteps
    - 10000000.0
  - - policy
    - CnnPolicy
  - - policy_kwargs
    - dict(optimizer_class=RMSpropTFLike, optimizer_kwargs=dict(eps=1e-5))
  - - vf_coef
    - 0.25
```
args.yml:
```
!!python/object/apply:collections.OrderedDict
- - - algo
    - a2c
  - - env
    - BreakoutNoFrameskip-v4
  - - env_kwargs
    - null
  - - eval_episodes
    - 5
  - - eval_freq
    - 10000
  - - gym_packages
    - []
  - - hyperparams
    - null
  - - log_folder
    - logs
  - - log_interval
    - -1
  - - n_eval_envs
    - 1
  - - n_evaluations
    - 20
  - - n_jobs
    - 1
  - - n_startup_trials
    - 10
  - - n_timesteps
    - -1
  - - n_trials
    - 10
  - - no_optim_plots
    - false
  - - num_threads
    - -1
  - - optimization_log_path
    - null
  - - optimize_hyperparameters
    - false
  - - pruner
    - median
  - - sampler
    - tpe
  - - save_freq
    - -1
  - - save_replay_buffer
    - false
  - - seed
    - 0
  - - storage
    - null
  - - study_name
    - null
  - - tensorboard_log
    - ''
  - - trained_agent
    - ''
  - - truncate_last_trajectory
    - true
  - - uuid
    - false
  - - vec_env
    - dummy
  - - verbose
    - 1
```

**Comment by qgallouedec** [comment]
Have you updated ˋgardient_steps` properly?

See related issue: https://github.com/DLR-RM/stable-baselines3/issues/699

Ans the documentation: https://stable-baselines3.readthedocs.io/en/master/guide/examples.html#multiprocessing-with-off-policy-algorithms

**Comment by Darklanx** [comment]
a2c is an on-policy algorithm, I think this is a different scenario?

**Comment by qgallouedec** [comment]
Indeed

**Comment by Darklanx** [comment]
Just to make sure, the observed behavior of `n_envs` having huge impact is unwanted, and when using hyperparameter tuning, tuning `n_envs` does not make sense, is this correct? If that's the case, I think this issue is actually more of a bug then a question?

**Comment by nmnaughton** [comment]
I've dealt with a similar issue to this with PPO. `n_envs` spawns that number of environments that each run for `n_steps`, in which case, increasing `n_envs` alone effectively increases your batch size, which can have a large effect on performance since you are updating 2.5x less often. Try decreasing `n_steps` by ~2.5x to see if that recovers performance? 

From the [a2c docs](https://stable-baselines3.readthedocs.io/en/master/modules/a2c.html#parameters): 
> n_steps (int) – The number of steps to run for each environment per update (i.e. batch size is n_steps * n_env where n_env is number of environment copies running in parallel)

**Comment by araffin** [comment]
> I thought n_envs is more of a parallelization parameter, which shouldn't have such a huge impact on performance.
> n_envs spawns that number of environments that each run for n_steps, in which case, increasing n_envs alone effectively increases your batch size

@nmnaughton is right, what's happening is that you are increasing the batch size without changing any other parameters, so a decrease in learning speed is expected.
What you can try to compensate is increasing the learning rate (but this won't be completely equivalent).
If you want to learn more about the compromise between learning speed and sample efficiency, you can take a look at our "Multiprocessing" notebook here: https://github.com/araffin/rl-tutorial-jnrr19#content

**Comment by Darklanx** [comment]
Thanks for clarifying, it now make sense.
